File: packages/contracts/python/demodirector_contracts/motion.py
from __future__ import annotations

import re
from datetime import datetime
from typing import Literal

from pydantic import Field, field_validator, model_validator

from demodirector_contracts.editorial import EditorialTemplateDraft, EditorialTemplatePlan
from demodirector_contracts.models import ContractModel, NonEmptyString

MotionPrimitive = Literal[
    "fade_slide",
    "scale_settle",
    "stagger_text",
    "highlight_reveal",
    "browser_frame_move",
    "background_dim",
]
MotionLayer = Literal[
    "background",
    "product",
    "mask",
    "cursor",
    "callout",
    "captions",
    "transition",
]
MotionEasing = Literal["linear", "standard", "emphasized"]
MotionDesignVersion = Literal["motion-v1"]
MOTION_LAYER_ORDER: tuple[MotionLayer, ...] = (
    "background",
    "product",
    "mask",
    "cursor",
    "callout",
    "captions",
    "transition",
)
UNSAFE_GENERATED_TEXT = re.compile(
    r"(?i)(<\s*script|javascript\s*:|data\s*:\s*text/html|\$\(|`[^`]*`|(?:^|\s)(?:rm|del)\s+-[rf])"
)


class MotionTokenSet(ContractModel):
    version: MotionDesignVersion = "motion-v1"
    font_family: Literal["Inter"] = "Inter"
    title_size: int = Field(default=64, ge=32, le=96)
    body_size: int = Field(default=32, ge=18, le=48)
    fast_ms: int = Field(default=200, ge=100, le=400)
    normal_ms: int = Field(default=600, ge=300, le=900)
    deliberate_ms: int = Field(default=1200, ge=700, le=1800)
    standard_easing: Literal["cubic-bezier(0.22, 1, 0.36, 1)"] = (
        "cubic-bezier(0.22, 1, 0.36, 1)"
    )
    emphasized_easing: Literal["cubic-bezier(0.16, 1, 0.3, 1)"] = (
        "cubic-bezier(0.16, 1, 0.3, 1)"
    )
    background: Literal["#111214"] = "#111214"
    surface: Literal["#1B1D20"] = "#1B1D20"
    text: Literal["#F5F6F7"] = "#F5F6F7"
    accent: Literal["#86E1A8"] = "#86E1A8"
    layer_order: list[MotionLayer] = Field(default_factory=lambda: list(MOTION_LAYER_ORDER))

    @model_validator(mode="after")
    def tokens_are_ordered(self) -> MotionTokenSet:
        if not self.fast_ms < self.normal_ms < self.deliberate_ms:
            raise ValueError("Motion durations must increase from fast to deliberate")
        if tuple(self.layer_order) != MOTION_LAYER_ORDER:
            raise ValueError("Motion layer order is fixed for deterministic composition")
        return self


class MotionCueDraft(ContractModel):
    scene_id: NonEmptyString
    primitive: MotionPrimitive
    layer: MotionLayer
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)
    easing: MotionEasing
    text: str | None = Field(default=None, max_length=120)
    product_clip_ref: str = Field(pattern=r"^scene:[A-Za-z0-9._:-]+$")

    @field_validator("text")
    @classmethod
    def text_is_content_only(cls, value: str | None) -> str | None:
        if value is not None and UNSAFE_GENERATED_TEXT.search(value):
            raise ValueError("Motion text cannot contain executable content")
        return value

    @model_validator(mode="after")
    def interval_and_text_are_valid(self) -> MotionCueDraft:
        if self.end_ms <= self.start_ms:
            raise ValueError("Motion cue end must follow its start")
        if self.primitive in {"stagger_text", "highlight_reveal"} and not self.text:
            raise ValueError("Text motion primitives require text")
        return self


class MotionCue(MotionCueDraft):
    id: NonEmptyString


class MotionDirectionDraft(ContractModel):
    summary: str = Field(min_length=1, max_length=240)
    cues: list[MotionCueDraft] = Field(min_length=1, max_length=80)
    editorial: EditorialTemplateDraft | None = None


class MotionDirectionRequest(ContractModel):
    project_id: NonEmptyString
    job_id: NonEmptyString
    storyboard_id: NonEmptyString
    storyboard_version: int = Field(gt=0)
    duration_ms: int = Field(gt=0, le=600_000)
    evidence_fingerprint: str = Field(pattern=r"^[a-f0-9]{64}$")
    source_ids: list[NonEmptyString] = Field(max_length=100)
    scene_ids: list[NonEmptyString] = Field(min_length=1, max_length=80)
    product_clip_refs: list[str] = Field(min_length=1, max_length=80)
    created_at: datetime

    @model_validator(mode="after")
    def references_are_unique_and_bounded(self) -> MotionDirectionRequest:
        for values, label in (
            (self.source_ids, "source"),
            (self.scene_ids, "scene"),
            (self.product_clip_refs, "product clip"),
        ):
            if len(values) != len(set(values)):
                raise ValueError(f"Motion request {label} references must be unique")
        expected_clips = {f"scene:{scene_id}" for scene_id in self.scene_ids}
        if set(self.product_clip_refs) - expected_clips:
            raise ValueError("Product clip references must identify requested scenes")
        return self


class MotionDirectionPlan(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    job_id: NonEmptyString
    run_id: NonEmptyString
    request_fingerprint: str = Field(pattern=r"^[a-f0-9]{64}$")
    design_tokens: MotionTokenSet
    duration_ms: int = Field(gt=0, le=600_000)
    scene_ids: list[NonEmptyString] = Field(min_length=1, max_length=80)
    product_clip_refs: list[str] = Field(min_length=1, max_length=80)
    summary: str = Field(min_length=1, max_length=240)
    cues: list[MotionCue] = Field(min_length=1, max_length=80)
    editorial_plan: EditorialTemplatePlan | None = None
    created_at: datetime

    @model_validator(mode="after")
    def cues_are_owned_ordered_and_non_overlapping(self) -> MotionDirectionPlan:
        if len(self.scene_ids) != len(set(self.scene_ids)):
            raise ValueError("Motion plan scene references must be unique")
        if len(self.product_clip_refs) != len(set(self.product_clip_refs)):
            raise ValueError("Motion plan product clip references must be unique")
        ids = [cue.id for cue in self.cues]
        if len(ids) != len(set(ids)):
            raise ValueError("Motion cue IDs must be unique")
        scenes = set(self.scene_ids)
        clips = set(self.product_clip_refs)
        if any(cue.scene_id not in scenes for cue in self.cues):
            raise ValueError("Motion cues may reference planned scenes only")
        if any(cue.product_clip_ref not in clips for cue in self.cues):
            raise ValueError("Motion cues may reference planned product clips only")
        if any(cue.end_ms > self.duration_ms for cue in self.cues):
            raise ValueError("Motion cues must fit the planned duration")
        ordered = sorted(self.cues, key=lambda cue: (cue.layer, cue.start_ms, cue.end_ms))
        previous_by_layer: dict[MotionLayer, MotionCue] = {}
        for cue in ordered:
            previous = previous_by_layer.get(cue.layer)
            if previous is not None and cue.start_ms < previous.end_ms:
                raise ValueError("Motion cues on the same layer cannot overlap")
            previous_by_layer[cue.layer] = cue
        return self


class ADKMotionRun(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    job_id: NonEmptyString
    session_id: NonEmptyString
    agent_name: Literal["demodirector_motion_director"]
    model: NonEmptyString
    status: Literal["running", "succeeded", "failed"]
    started_at: datetime
    completed_at: datetime | None = None
    elapsed_ms: int | None = Field(default=None, ge=0)
    retry_count: int = Field(ge=0, le=2)
    workflow_runs: int = Field(ge=0, le=1)
    input_artifact_ids: list[NonEmptyString] = Field(min_length=1, max_length=120)
    output_plan_id: NonEmptyString | None = None
    validation_status: Literal["pending", "passed", "failed"]
    message: str = Field(min_length=1, max_length=300)

    @model_validator(mode="after")
    def execution_state_is_consistent(self) -> ADKMotionRun:
        if self.status == "running":
            if self.completed_at is not None or self.elapsed_ms is not None:
                raise ValueError("Running ADK motion runs cannot have completion timing")
            if self.validation_status != "pending" or self.output_plan_id is not None:
                raise ValueError("Running ADK motion runs cannot have validated output")
        else:
            if self.completed_at is None or self.elapsed_ms is None:
                raise ValueError("Completed ADK motion runs require timing")
            expected = max(
                0,
                round((self.completed_at - self.started_at).total_seconds() * 1000),
            )
            if abs(expected - self.elapsed_ms) > 5:
                raise ValueError("ADK motion elapsed time must match its timestamps")
        if self.status == "succeeded" and (
            self.workflow_runs != 1
            or self.validation_status != "passed"
            or self.output_plan_id is None
        ):
            raise ValueError("Successful ADK motion runs require one validated output")
        if self.status == "failed" and (
            self.validation_status != "failed" or self.output_plan_id is not None
        ):
            raise ValueError("Failed ADK motion runs cannot expose an output plan")
        unsafe = self.message.casefold()
        if any(marker in unsafe for marker in ("api_key=", "bearer ", "password=", "secret=")):
            raise ValueError("ADK motion messages cannot contain credentials")
        return self


class CompiledMotionCue(ContractModel):
    cue_id: NonEmptyString
    primitive: MotionPrimitive
    layer: MotionLayer
    layer_index: int = Field(ge=0, le=6)
    start_frame: int = Field(ge=0)
    end_frame: int = Field(gt=0)
    easing: MotionEasing


class CompiledMotionComposition(ContractModel):
    project_id: NonEmptyString
    plan_id: NonEmptyString
    design_version: Literal["motion-v1"]
    width: int = Field(ge=320, le=3840)
    height: int = Field(ge=180, le=2160)
    fps: int = Field(ge=12, le=60)
    duration_ms: int = Field(gt=0)
    cues: list[CompiledMotionCue]
    deterministic_hash: str = Field(pattern=r"^[a-f0-9]{64}$")

File: packages/contracts/src/motion.ts
import { z } from "zod";
import { editorialTemplateDraftSchema, editorialTemplatePlanSchema } from "./editorial";

export const motionPrimitiveSchema = z.enum(["fade_slide", "scale_settle", "stagger_text", "highlight_reveal", "browser_frame_move", "background_dim"]);
export const motionLayerSchema = z.enum(["background", "product", "mask", "cursor", "callout", "captions", "transition"]);
export const motionEasingSchema = z.enum(["linear", "standard", "emphasized"]);
export const motionDesignVersionSchema = z.literal("motion-v1");
const layerOrder = ["background", "product", "mask", "cursor", "callout", "captions", "transition"] as const;
const unsafeText = /<\s*script|javascript\s*:|data\s*:\s*text\/html|\$\(|`[^`]*`|(?:^|\s)(?:rm|del)\s+-[rf]/i;

export const motionTokenSetSchema = z.object({
  version: z.literal("motion-v1").default("motion-v1"), font_family: z.literal("Inter").default("Inter"),
  title_size: z.number().int().min(32).max(96).default(64), body_size: z.number().int().min(18).max(48).default(32),
  fast_ms: z.number().int().min(100).max(400).default(200), normal_ms: z.number().int().min(300).max(900).default(600), deliberate_ms: z.number().int().min(700).max(1800).default(1200),
  standard_easing: z.literal("cubic-bezier(0.22, 1, 0.36, 1)").default("cubic-bezier(0.22, 1, 0.36, 1)"),
  emphasized_easing: z.literal("cubic-bezier(0.16, 1, 0.3, 1)").default("cubic-bezier(0.16, 1, 0.3, 1)"),
  background: z.literal("#111214").default("#111214"), surface: z.literal("#1B1D20").default("#1B1D20"),
  text: z.literal("#F5F6F7").default("#F5F6F7"), accent: z.literal("#86E1A8").default("#86E1A8"),
  layer_order: z.array(motionLayerSchema).default([...layerOrder]),
}).strict().superRefine((tokens, ctx) => {
  if (!(tokens.fast_ms < tokens.normal_ms && tokens.normal_ms < tokens.deliberate_ms)) ctx.addIssue({ code: "custom", message: "Motion durations must be ordered" });
  if (tokens.layer_order.join("|") !== layerOrder.join("|")) ctx.addIssue({ code: "custom", message: "Motion layer order is fixed" });
});

const motionCueDraftFields = {
  scene_id: z.string().min(1), primitive: motionPrimitiveSchema, layer: motionLayerSchema,
  start_ms: z.number().int().nonnegative(), end_ms: z.number().int().positive(), easing: motionEasingSchema,
  text: z.string().max(120).nullable(), product_clip_ref: z.string().regex(/^scene:[A-Za-z0-9._:-]+$/),
};
const motionPrimitiveBase = z.object(motionCueDraftFields).strict();
const refineCue = (cue: z.infer<typeof motionPrimitiveBase>, ctx: z.RefinementCtx) => {
  if (cue.end_ms <= cue.start_ms) ctx.addIssue({ code: "custom", message: "Motion cue end must follow start" });
  if (["stagger_text", "highlight_reveal"].includes(cue.primitive) && !cue.text) ctx.addIssue({ code: "custom", message: "Text primitive requires text" });
  if (cue.text && unsafeText.test(cue.text)) ctx.addIssue({ code: "custom", message: "Motion text cannot contain executable content" });
};
export const motionCueDraftSchema = motionPrimitiveBase.superRefine(refineCue);
export const motionCueSchema = z.object({ id: z.string().min(1), ...motionCueDraftFields }).strict().superRefine(refineCue);
export const motionDirectionDraftSchema = z.object({
  summary: z.string().min(1).max(240),
  cues: z.array(motionCueDraftSchema).min(1).max(80),
  editorial: editorialTemplateDraftSchema.nullable().default(null),
}).strict();
export const motionDirectionRequestSchema = z.object({
  project_id: z.string().min(1), job_id: z.string().min(1), storyboard_id: z.string().min(1), storyboard_version: z.number().int().positive(),
  duration_ms: z.number().int().positive().max(600000), evidence_fingerprint: z.string().regex(/^[a-f0-9]{64}$/),
  source_ids: z.array(z.string().min(1)).max(100), scene_ids: z.array(z.string().min(1)).min(1).max(80),
  product_clip_refs: z.array(z.string().regex(/^scene:[A-Za-z0-9._:-]+$/)).min(1).max(80), created_at: z.string().datetime({ offset: true }),
}).strict().superRefine((request, ctx) => {
  const unique = (values: string[]) => new Set(values).size === values.length;
  if (!unique(request.source_ids) || !unique(request.scene_ids) || !unique(request.product_clip_refs)) ctx.addIssue({ code: "custom", message: "Motion request references must be unique" });
  const expected = new Set(request.scene_ids.map((id) => `scene:${id}`));
  if (request.product_clip_refs.some((ref) => !expected.has(ref))) ctx.addIssue({ code: "custom", message: "Product clip must identify a requested scene" });
});
export const motionDirectionPlanSchema = z.object({
  id: z.string().min(1), project_id: z.string().min(1), job_id: z.string().min(1), run_id: z.string().min(1),
  request_fingerprint: z.string().regex(/^[a-f0-9]{64}$/), design_tokens: motionTokenSetSchema,
  duration_ms: z.number().int().positive().max(600000), scene_ids: z.array(z.string().min(1)).min(1).max(80),
  product_clip_refs: z.array(z.string()).min(1).max(80), summary: z.string().min(1).max(240),
  cues: z.array(motionCueSchema).min(1).max(80), created_at: z.string().datetime({ offset: true }),
  editorial_plan: editorialTemplatePlanSchema.nullable().default(null),
}).strict().superRefine((plan, ctx) => {
  const fail = (message: string) => ctx.addIssue({ code: "custom", message });
  if (new Set(plan.scene_ids).size !== plan.scene_ids.length || new Set(plan.product_clip_refs).size !== plan.product_clip_refs.length || new Set(plan.cues.map((cue) => cue.id)).size !== plan.cues.length) fail("Motion references must be unique");
  const scenes = new Set(plan.scene_ids); const clips = new Set(plan.product_clip_refs);
  if (plan.cues.some((cue) => !scenes.has(cue.scene_id) || !clips.has(cue.product_clip_ref) || cue.end_ms > plan.duration_ms)) fail("Motion cues must reference owned inputs and fit duration");
  for (const layer of layerOrder) {
    const cues = plan.cues.filter((cue) => cue.layer === layer).sort((a, b) => a.start_ms - b.start_ms);
    if (cues.some((cue, index) => index > 0 && cue.start_ms < cues[index - 1].end_ms)) fail("Motion cues on one layer cannot overlap");
  }
});
export const adkMotionRunSchema = z.object({
  id: z.string().min(1), project_id: z.string().min(1), job_id: z.string().min(1), session_id: z.string().min(1),
  agent_name: z.literal("demodirector_motion_director"), model: z.string().min(1), status: z.enum(["running", "succeeded", "failed"]),
  started_at: z.string().datetime({ offset: true }), completed_at: z.string().datetime({ offset: true }).nullable(), elapsed_ms: z.number().int().nonnegative().nullable(),
  retry_count: z.number().int().min(0).max(2), workflow_runs: z.number().int().min(0).max(1), input_artifact_ids: z.array(z.string().min(1)).min(1).max(120),
  output_plan_id: z.string().min(1).nullable(), validation_status: z.enum(["pending", "passed", "failed"]), message: z.string().min(1).max(300),
}).strict().superRefine((run, ctx) => {
  const fail = (message: string) => ctx.addIssue({ code: "custom", message });
  if (run.status === "running" && (run.completed_at !== null || run.elapsed_ms !== null || run.validation_status !== "pending" || run.output_plan_id !== null)) fail("Running ADK motion state is invalid");
  if (run.status !== "running") {
    if (!run.completed_at || run.elapsed_ms === null) fail("Completed ADK motion run requires timing");
    else if (Math.abs(new Date(run.completed_at).getTime() - new Date(run.started_at).getTime() - run.elapsed_ms) > 5) fail("ADK timing must match");
  }
  if (run.status === "succeeded" && (run.workflow_runs !== 1 || run.validation_status !== "passed" || !run.output_plan_id)) fail("Successful ADK motion run requires validated output");
  if (run.status === "failed" && (run.validation_status !== "failed" || run.output_plan_id !== null)) fail("Failed ADK motion run cannot expose output");
  if (/api_key=|bearer |password=|secret=/i.test(run.message)) fail("ADK motion message contains credentials");
});
export const compiledMotionCueSchema = z.object({
  cue_id: z.string().min(1), primitive: motionPrimitiveSchema, layer: motionLayerSchema, layer_index: z.number().int().min(0).max(6),
  start_frame: z.number().int().nonnegative(), end_frame: z.number().int().positive(), easing: motionEasingSchema,
}).strict();
export const compiledMotionCompositionSchema = z.object({
  project_id: z.string().min(1), plan_id: z.string().min(1), design_version: z.literal("motion-v1"),
  width: z.number().int().min(320).max(3840), height: z.number().int().min(180).max(2160), fps: z.number().int().min(12).max(60), duration_ms: z.number().int().positive(),
  cues: z.array(compiledMotionCueSchema), deterministic_hash: z.string().regex(/^[a-f0-9]{64}$/),
}).strict();

export type ADKMotionRun = z.infer<typeof adkMotionRunSchema>;
export type MotionDesignVersion = z.infer<typeof motionDesignVersionSchema>;
export type MotionDirectionPlan = z.infer<typeof motionDirectionPlanSchema>;
export type MotionDirectionRequest = z.infer<typeof motionDirectionRequestSchema>;
